octopus: update to 11.4 and add tests

fangohr opened the pull request:
- add more versions of Octopus (including the most recent 11.4 at the moment)
- add smoke tests
- add ability to build from git repository head rather than released versions
- adding more variants
- add required flags for compilation with gcc@10

fangohr committed:
Merge pull request #4 from spack/develop

fetch upstream

fangohr committed:
Merge branch 'spack:develop' into develop

fangohr committed:
octopus: adding versions up to 11.4

fangohr committed:
octopus: add smoke tests

fangohr committed:
octopus: add necessary flags for gcc@10

fangohr committed:
octopus: update to compilation and dependencies

fangohr committed:
octopus: adding new variants

fangohr committed:
octopus: allow compilation from git repo develop branch

fangohr committed:
octopus: adapt to spack style requirements

fangohr committed:
octopus: remove 'poke' (as this poke is not in spack [yet])

fangohr committed:
octopus: add maintainer

tldahlgren commented on var/spack/repos/builtin/packages/octopus/package.py:
```diff
@@ -134,12 +228,99 @@ def install(self, spec, prefix):
+        self.smoke_tests()
+
+    @run_after("install")
+    def smoke_tests(self):
```
Have you tested this using `spack test run`?

IIRC The `run_after` directive will cause problems with that use case.

fangohr replied:
I didn't get the pattern to use the `spack install --test=root` flag to run tests during install; thank you for pointing this out below. With this new knowledge, I do indeed not want to run the tests automatically during install. 

However, I would like to re-use the same testing code during install, and also when `spack test run` is used. To avoid duplication, I have created three functions:


```python
 @run_after('install')
    @on_package_attributes(run_tests=True)
    def smoke_tests_after_install(self):
        """Function stub to run tests after install if desired
        (for example through `spack install --test=root octopus`)
        """
        self.smoke_tests()

    def test(self):
        """Entry point for smoke tests run through `spack test run octopus`.
        """
        self.smoke_tests()

    def smoke_tests(self):
        """Actual smoke tests for Octopus."""
        #
```

I have tested this, and believe it works as expected.

tldahlgren commented on var/spack/repos/builtin/packages/octopus/package.py:
```diff
@@ -134,12 +228,99 @@ def install(self, spec, prefix):
+
+    @run_after("install")
+    def smoke_tests(self):
+        """Run these smoke tests when requested explicitly"""
```
What do you mean in this case?  

Your use of the `run_after` directive means it will always run after the `install` phase of `spack install`.

If you want it to be optionally run after the installation (e.g., `spack install --test=root`) then you need to add the `on_package_attributes` directory. 

You can find more information on testing a package at https://spack.readthedocs.io/en/latest/packaging_guide.html#checking-an-installation.

fangohr replied:
Thanks - this is now addressed in the discussion above (https://github.com/spack/spack/pull/30321#discussion_r861241496).

tldahlgren reviewed (commented):
Chiming in on the test aspects of this PR.

I would generally recommend putting the `*.inp` files in a `test` subdirectory.  These would automatically be copied to the test stage directory for you by `spack test run`.    (See https://spack.readthedocs.io/en/latest/packaging_guide.html#adding-custom-files.)

You would, however, need to copy them yourself if you want to re-use the method with `spack install`.

fangohr committed:
octopus: make tests after install optional

Thank you @tldahlgren

fangohr committed:
octopus: follow recommended practice for test input data

Move the two configuration files we use for smoke tests into `test`
subdirectory. Thanks @tldahlgren.

fangohr commented:
> Chiming in on the test aspects of this PR.
> 
> I would generally recommend putting the `*.inp` files in a `test` subdirectory. 

I wasn't aware of the practice - thank you for pointing this out and providing the link to the documentation. 

I have moved the `*.inp` files into the `test` directory now (2750f5829966f0997e5a0335d64b784932a4cc08).

> These would automatically be copied to the test stage directory for you by `spack test run`. (See 
> https://spack.readthedocs.io/en/latest/packaging_guide.html#adding-custom-files.)
That's good to know, thank you.

> 
> You would, however, need to copy them yourself if you want to re-use the method with `spack install`.
As I would like to be able to run the tests at install time, I have kept the 'manual' coping as before.

RemiLacroix-IDRIS commented on var/spack/repos/builtin/packages/octopus/package.py:
```diff
@@ -134,12 +228,113 @@ def install(self, spec, prefix):
+            # argument mismatching. Need to provide a flag to turn this into a
+            # warning and build sucessfully
+            if (spec.satisfies('%gcc@10:')):
+                args.extend([
```
I think you could avoid duplicating the flags that are common to all versions.
```
args.extends(['common_flags'])
if ...:
  args.extends(['gcc10_specific_flags'])
```

fangohr replied:
I have addressed that in https://github.com/spack/spack/pull/30321/files/a4dfd5f4f42c0217df7669cda4ea7a4f69faaef1..255d665fae6a686e4ea51f3d2f9c54f55638ef09 .

fangohr committed:
Adding maintainer

with their agreement by email

fangohr committed:
octopus: reduce duplication of flags

- part of code review

tldahlgren reviewed (commented):
The test inputs copy seems reasonable under the circumstances.

RemiLacroix-IDRIS commented on var/spack/repos/builtin/packages/octopus/package.py:
```diff
@@ -31,26 +45,62 @@ class Octopus(Package, CudaPackage):
-    depends_on('libxc@2.0:3', when='@6:7')
-    depends_on('libxc@2.0:4', when='@8:9')
-    depends_on('libxc@3:5.0.0', when='@10:')
+    depends_on('libxc@2:2.99', when='@:5.99')
```
I think you don't need the `.99` in the version ranges now.

fangohr replied:
So I could replace line 78 with `depends_on('libxc@2:2.99', when='@:5')` ?

RemiLacroix-IDRIS replied:
See: https://github.com/spack/spack/commit/b9e72557e8b52b63244500f226afda0efe798df2#diff-b8373d30b3a141c495c2281273ee6184fc513413142afaf2adac1f406cd6b0d7

fangohr replied:
Thank you. I have implemented this (https://github.com/spack/spack/pull/30321/commits/492d0ce8668d98de9db59352a3347c51b2a49ed8).

RemiLacroix-IDRIS commented on var/spack/repos/builtin/packages/octopus/package.py:
```diff
@@ -65,6 +115,7 @@ def install(self, spec, prefix):
             'CC=%s' % spec['mpi'].mpicc,
             'FC=%s' % spec['mpi'].mpifc,
             '--enable-mpi',
```
Would it make sense to have MPI and OpenMP as variants (not sure myself, just wondering)?

fangohr replied:
That is a good question. I'll find out (Octopus developers are nearby) but would suggest to leave this for a later pull request. (I expect more modifications to the file in any case).

tldahlgren commented on var/spack/repos/builtin/packages/octopus/package.py:
```diff
@@ -114,11 +165,54 @@ def install(self, spec, prefix):
             ])
 
+        if '+cgal' in spec:
+            args.extend([
+                '--with-cgal-prefix=%s' % spec['cgal'].prefix,
+            ])
```
FWIW.  You could use `append` when only one option/string is being added to the args:

```suggestion
            args.append(
                '--with-cgal-prefix=%s' % spec['cgal'].prefix
            )
```

fangohr replied:
Good point - let's keep it as simple as possible. That's implemented in https://github.com/spack/spack/pull/30321/commits/b9035495716dbb0abb672a478b645e1c970b4c9b .

tldahlgren commented on var/spack/repos/builtin/packages/octopus/package.py:
```diff
@@ -14,13 +16,25 @@ class Octopus(Package, CudaPackage):
 
     homepage = "https://octopus-code.org/"
     url      = "http://octopus-code.org/down.php?file=6.0/octopus-6.0.tar.gz"
```
FYI. We prefer `https` over `http`. I made the change and `spack checksum` works, though it did take a long time for Spack to fetch the new versions with either protocol.

fangohr replied:
Okay, I implemented the change. (https://github.com/spack/spack/pull/30321/commits/d673503feae8905727effe714c57f3a423f96922)

tldahlgren reviewed (dismissed):
Confirmed sha256.

tldahlgren commented:
Let me know if this is ready to merge or if you want or need to make a few tweaks.

fangohr committed:
octopus: https is preferred over http

fangohr committed:
octopus: remove .99 from versioning information

Thanks to https://github.com/spack/spack/pull/26402, we can drop the
"2:3.99" notation when we mean all versions 2.x and 3.x

Examples: https://github.com/spack/spack/commit/b9e72557e8b52b63244500f226afda0efe798df2#diff-b8373d30b3a141c495c2281273ee6184fc513413142afaf2adac1f406cd6b0d7

(from review)

fangohr committed:
octopus: args.extend([x]) -> args.append(x)

(hint from review)

fangohr commented:
> Let me know if this is ready to merge or if you want or need to make a few tweaks.

I am happy for this to be merged (thanks for waiting - I have now implemented most of the suggestions made in the last 24 hours).

Thank you @tldahlgren and @RemiLacroix-IDRIS for your input and help!
